test/integration/proxy_test.go

```diff
@@ -0,0 +1,105 @@
+// +build integration
+
+/*
+Copyright 2019 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"net/http"
+
+	"github.com/elazarl/goproxy"
+	"github.com/phayes/freeport"
+	"github.com/pkg/errors"
+)
+
+// setUpProxy runs a local http proxy and sets the env vars for it.
+func setUpProxy(t *testing.T) (*http.Server, error) {
+	port, err := freeport.GetFreePort()
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to get an open port")
+	}
+
+	addr := fmt.Sprintf("localhost:%d", port)
+	err = os.Setenv("NO_PROXY", "")
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to set no proxy env")
+	}
+	err = os.Setenv("HTTP_PROXY", addr)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to set http proxy env")
+	}
+
+	proxy := goproxy.NewProxyHttpServer()
+	srv := &http.Server{Addr: addr, Handler: proxy}
+	go func(s *http.Server, t *testing.T) {
+		if err := s.ListenAndServe(); err != http.ErrServerClosed {
+			t.Errorf("Failed to start http server for proxy mock")
+		}
+	}(srv, t)
+	return srv, nil
+}
+
+func TestProxy(t *testing.T) {
+	srv, err := setUpProxy(t)
```

Comment:
Consider adding: `defer os.Setenv("HTTP_PROXY", "")` so that other tests are not affected.